Updating the HWIntrinsic APIs to use the unmanaged constraint

tannergooding opened the pull request:
FYI. @CarolEidt, @fiigii, @sdmaclea, @eerhardt, @jkotas 

This is mostly just for discussion right now, since the actual change can't be merged until after the branch opens again.

This is a new (and unshipped API) so back-compat should not be a concern here. However, there are some potential other issues with this change:
* The `unmanaged` constraint uses a `modreq` on the constraint, there are several known places where tools won't respect this
* The `modreq` also prevents compiler versions prior to C# 7.3 from consuming these APIs.

Some of the benefits are:
* The constraint is more accurate (but still not perfect)
* You can use unsafe operations with these types
  * This includes both unsafe operations that can and cannot be done with `System.Runtime.CompilerServices.Unsafe`
  * Examples of things not possible with a shim API include stackalloc and bypassing definite assignment rules

jkotas commented:
> Some of the benefits

Note that there are potential disadvantages too. For example:

```
static T[] AddAnything<T>(T[] x, T[] y)
{
    if (Simd.IsSupported && IsSupportedBySimd<T>())
    {
        // You can use `Simd.Add<T>` here today, but you won't be able to use it with your proposed change. You will have to
        // manually expand the code for every T
    }
    else
    {
        ... other cases ...
    }
}

static bool IsSupportedBySimd<T>() => typeof(T) == typeof(float) || typeof(T) == typeof(double) || typeof(T) == ...;
```

> Examples of things not possible with a shim API include stackalloc and bypassing definite assignment rules

You can still do these using `Unsafe` (stackalloc byte array of the right size and cast it, etc.). It is not pretty, but it works.

tannergooding commented:
> Note that there are potential disadvantages too. For example:

Right, you would need to constrain `T` for anything that tries to use it. A good question might be if this type of code represents "real-world" use-cases.

> You can still do these using Unsafe (stackalloc byte array of the right size and cast it, etc.). It is not pretty, but it works.

You can get stackalloc to work, with casting, but bypassing definite assignment isn't possible without C# saying it is okay to take the address of the containing struct.

Ex:
```C#
using System.Runtime.InteropServices;

[StructLayout(LayoutKind.Explicit)]
unsafe struct VectorF32
{
    #region union
    [FieldOffset(0)]
    public fixed float f[4];
    
    [FieldOffset(0)]
    public Vector128<float> v;
    #endregion

    public VectorF32(float x, float y, float z, float w)
    {
        fixed (VectorF32* pThis = &this)
        {
            // This bypasses C# definite assignment rules and allows
            // for more efficient codegen. It is currently optimized
            // away by the JIT.
        }

        // Don't need to take the address of fixed buffers in 7.3
        f[0] = x;
        f[1] = y;
        f[2] = z;
        f[3] = w;
    }

    public VectorF32(Vector128<float> value)
    {
        fixed (VectorF32* pThis = &this)
        {
            // This bypasses C# definite assignment rules and allows
            // for more efficient codegen. It is currently optimized
            // away by the JIT.
        }

        v = value;
    }
}
```

jkotas commented:
> A good question might be if this type of code represents "real-world" use-cases.

FWIW, the same convern was raised in https://github.com/dotnet/corefx/issues/29119#issuecomment-381429090

> `public fixed float f[4]`

This field is going to flag the struct as prone to stack buffer overruns (...), that is in turn going to generate GS cookie checks in all methods that use it. You do not want to use this construct when low-level performance is important.

tannergooding commented:
I opened https://github.com/dotnet/csharplang/issues/1465 as a possible alternative.

tannergooding commented:
Going to close for right now. It actually doesn't look possible to do `Vector128<T>* pV = &v` anymore.
